text/0000-offset_of.md

```diff
@@ -0,0 +1,576 @@
+possibilities.
+
+## Nested Field Access
+
+In C, expressions like `offsetof(struct some_struct, foo.bar.baz[3].quux)` are
+allowed, where `foo.bar.baz[3].quux` denotes a path to a derived field. This can
+be of somewhat arbitrary complexity, accessing fields of nested structs,
+performing array indexing (often this is used to access past the end of the
+array even), and so on. Similar functionality is offered by
+`MemoryLayout.offset` in Swift, where more complex language features are used to
+achieve it.
+
+This was omitted from this proposal because it is not commonly used, and can
+generally be replaced (at the cost of convenience) by multiple invocations of
+the macro.
+
+Additionally, in the future similar functionality could be added in a
+backwards-compatible way, either by directly allowing usage like
+`offset_of!(SomeStruct, foo.bar.baz[3].quux)`, or by requiring each field be
+comma-separated, as in `offset_of!(SomeStruct, foo, bar, baz, [3], quux)`.
+
+Note that while this example shows a combination that supports array indexing,
+it's unclear if this is actually desirable for Rust.
+
+## Enum support (`offset_of!(SomeEnum::StructVariant, field_on_variant)`)
+
+Eventually, it may be desirable to allow `offset_of!` to access the fields
+inside the struct and tuple variants of certain enums (possibly limited to enums
+with a primitive integer representation, such as `#[repr(C)]`, `#[repr(int)]`,
+or `#[repr(C, int)]` -- where `int` is one of Rust's primitive integer types —
+u8, isize, u128, etc).
+
+For example, in the future something like the following could be allowed:
+
+```rs
+use core::mem::offset_of;
+
+#[repr(i8)]
+enum Event {
+    Key { pressed: bool, code: u32 },
+    Resize(u32, u32),
+}
+
+const EVENT_KEY_CODE: usize = offset_of!(Event::Key, code);
+const EVENT_KEY_PRESSED: usize = offset_of!(Event::Key, pressed);
+
+const EVENT_RESIZE_W: usize = offset_of!(Event::Resize, 0);
+const EVENT_RESIZE_H: usize = offset_of!(Event::Resize, 1);
+```
+
+In this example, the name/path of the variant is used as the first argument.
+While there are use-cases for this in low level FFI code (similar to the use
+cases for `#[repr(int)]` and `#[repr(C, int)]` enums), this may need further
+design work, and is left to the future.
+
+A drawback is that it is unclear how to support these types in the "Nested Field
+Access" proposed above, so in the future should we decide to support one of
+these, a decision may need to be made about the other.
+
+## `memoffset::span_of!` Functionality
+
+The `memoffset` crate has support for a [`span_of!`][spanof] macro (used like
+`memoffset::span_of!(SomeType, some_field)`), which expands to a `Range<usize>`
+indicating which bytes of `SomeType` are from the field `some_field`.
+
+The use case for this is more limited than that of `offset_of!`, so it was
+omitted from this proposal. That said, should this prove sufficiently useful, it
+would be simple to add a similar macro to `core::mem` in the future.
```

Comment:
Scrubbed GitHub for where this is used, and commented [here](https://github.com/Gilnaa/memoffset/issues/59#issuecomment-1273678497). (for admittedly different motives)

tl;dr it's not used all that much in practice

Reply:
There are a lot of cases where I've made something repr(C) just so you can hard-code the offset. Several others have said as much in this issue.